### ISSUE TITLE
`P.kwargs` is not assignable to `dict[str, object]`

### ISSUE THREAD

**Issue by jonasdedden** [astral-sh/ty#4477]
### Summary

This is more of a discussion about an *potential* inconsistency, not really a bug per se. Feel free to close if you think this isn't leading anywhere.

```python
def kw[**P, R](func: Callable[P, R], *args: P.args, **kwargs: P.kwargs) -> R:
    a: tuple[object, ...] = args    # ty: OK
    k: dict[str, object] = kwargs   # ty: error[invalid-assignment]
                                    #     `P@kw.kwargs` is not assignable to `dict[str, object]`
    return func(*args, **kwargs)
```

I know the typing spec does not state what `P.args`/`P.kwargs` are assignable to (see [python/typing discussion #1447](https://github.com/python/typing/discussions/1447): "no mechanism currently in the type system"), so this is somewhat spec-underspecified rather than a bug, if anything.

=> Still, one could argue that `ty` is inconsistent with itself (`args` OK, `kwargs` not) and: `mypy`, `pyright` and `pyrefly` accept both assignments.

### Version

0.0.78

**Comment by sharkdp** [comment]
I think the current behavior here is intentional and correct.

If we would allow the assignment `k: dict[str, object] = kwargs`, then you'd be able to modify `kwargs` through `k`. And we could not verify the subsequent `func(*args, **kwargs)` call. Imagine:

```py
def evil(k: dict[str, object]):
    k["non_existing_arg"] = "whatever"

def kw[**P, R](func: Callable[P, R], *args: P.args, **kwargs: P.kwargs) -> R:
    evil(kwargs)  # ty does not (and should not) allow this call
    return func(*args, **kwargs)
```

>  ty is inconsistent with itself (args OK, kwargs not)

No, the `a: tuple[object, ...] = args` case is different. You can't add inconsistent args through that `a` reference, since you can't append to a tuple, or remove elements from it.

You can use `Mapping[str, object]` to do the same for `kwargs`. This is fine:
```py
k: Mapping[str, object] = kwargs
```

> `mypy`, `pyright` and `pyrefly` accept both assignments.

This means they won't catch the `evil` problem above.

**Comment by jonasdedden** [comment]
Ah, true. The problem lies in the mutability of the type, as the following also is throwing an error:

```
$ uv run --with ty ty check /tmp/ty_test.py 
error[invalid-assignment]: Object of type `P@kw.kwargs` is not assignable to `MutableMapping[str, object]`
 --> /tmp/ty_test.py:6:38
  |
6 |     k: MutableMapping[str, object] = kwargs   # ty: error[invalid-assignment]
  |        ---------------------------   ^^^^^^ Incompatible value of type `P@kw.kwargs`
  |        |
  |        Declared type

Found 1 diagnostic
```

Thanks for clearing that up!

Still: Isn't the error here technically at the wrong place or isn't it throwing the wrong error message?

Currently it throws: "`P@kw.kwargs` is not assignable to `dict[str, object]`". That sounds to me like `ty` thinks this isn't a `dict` (which it *is*). I could have perfectly valid code that isn't using `func` with modified `k`, but just does some logic with `k` where it mutates its keys and/or values or put it into other functions that expect a bare `dict` as input.

The problem you're pointing out rather is an issue how `k` is **used**. I.e., the `evil` problem you stated, *will* become an issue with the `return func(*args, **kwargs)` line.

~So I think one *could* argue that the errors should instead be:~

```
def kw[**P, R](func: Callable[P, R], *args: P.args, **kwargs: P.kwargs) -> R:
    a: tuple[object, ...] = args    # ty: OK
    k: dict[str, object] = kwargs   # ty: ok
    return func(*args, **kwargs)  # error: calling `func` with untyped `tuple[object, ...]` & `dict[str, object]` (could contain unwanted or wrongly typed `args` or `kwargs`)
```

~Am I missing something?~

EDIT: Sidethought: If CPython worked in a way that `**kwargs` inside the function would be an instance of the new `frozendict`, this typing issue would be gone (but it would introduce a whole lot of breaking changes with existing code of course).

EDIT2: Ah sorry, had a brainfart here. Sorry, yes, `kwargs` shall not even be assignable anywhere to a mutable type, as this is the issue actually (as then, for example through `k`, the original `kwargs` can be mutated). The callsite argument that I just did is wrong, as it's actually already using the fully typed `**kwargs`, but just because it was assigned to a mutable type somewhere, the guarantees of it no longer hold true. Closing the issue.

**Comment by jonasdedden** [comment]
I wanted to know what other type checkers do in a similar example (mutating `kwargs` in a potentially breaking way because `func` may not expect this parameter or not this type):

```Python
from collections.abc import Callable

def kw[**P, R](func: Callable[P, R], *args: P.args, **kwargs: P.kwargs) -> R:
    kwargs["hui"] = "fui"
    return func(*args, **kwargs)
```

checker | tiers tested | result on snippet |
|---|---|---|
| `mypy` | default, `--strict` | pass, `EXIT:0` — `Success: no issues found` |
| `pyright` | `off`, `basic`, `standard`, `strict` via `pyrightconfig.json` | pass on all, `0 errors` |
| `basedpyright` | `off`, `basic`, `standard`, `recommended`, `all`, `strict` via `basedpyrightconfig.json` | pass on all, `0 errors` |
| `pyrefly` | `off`, `basic`, `legacy`, `default`, `strict`, `all` via `pyrefly check --preset` | pass on all, `INFO 0 errors` |
| `ty` | default `ty check` | **error, `EXIT:1`:**

`ty` indeed is the only checker who flags this:
```
error[invalid-assignment]: Invalid subscript assignment with key of type `Literal["hui"]`
and value of type `Literal["fui"]` on object of type `P@kw.kwargs`
 --> kw.py:4:5
  |
4 |     kwargs["hui"] = "fui"
  |     ^^^^^^^^^^^^^^^^^^^^^
Found 1 diagnostic
```
